python/pyarrow/table.pxi

```diff
@@ -1669,26 +1669,44 @@ def table(data, names=None, schema=None, metadata=None):
             "Expected pandas DataFrame, python dictionary or list of arrays")
 
 
-def concat_tables(tables):
+def concat_tables(tables, c_bool promote=False, MemoryPool memory_pool=None):
     """
-    Perform zero-copy concatenation of pyarrow.Table objects. Raises exception
-    if all of the Table schemas are not the same
+    Perform zero-copy concatenation of pyarrow.Table objects.
+
+    If promote==False, the schemas of all the Tables must be the same
+    (modulo the metadata), otherwise an exception will be raised. The result
+    Table will share the metadata with the first table.
+
+    If promote==True, columns of the same name will be concatenated.
+    They should be of the same type, or be of type NULL, in which case it will
+    be promoted to the type of other corresponding columns with null values
+    filled.  If a table is missing a particular field, null values of the
+    appropriate type will be generated to take the place of the missing field.
+    The new schema will share the metadata with the first table. Each field in
+    the new schema will share the metadata with the first table which has the
```

Comment:
Is this tested already on the C++ side? (the current python tests don't cover this aspect I think)

Reply:
Yes. The C++ tests are comprehensive so I only wrote a python test to cover the wiring.